File: brisa-src/Brisa/src/upnp/brisastatevariable.cpp
#include "brisastatevariable.h"
#include "brisaabstractservice.h"

namespace brisa {
namespace upnp {

StateVariable::StateVariable(QString sendEvents,
                                       QString name,
                                       QString datatype,
                                       QString defaultValue,
                                       QString maximum,
                                       QString minimum,
                                       QString step,
                                       QString multicast,
                                       QObject *parent) :
    QObject(parent)
{
    this->nextMulticastSeq = 0;
    this->_sendEvents = (sendEvents == "yes");
    this->_multicast = (multicast == "yes");
    this->_name = name;
    this->_dataType = datatype;
    this->_defaultValue = defaultValue;
    this->_maximum = maximum;
    this->_minimum = minimum;
    this->_step = step;
    this->_value = this->defaultValue;
}

StateVariable::StateVariable(const StateVariable &variable) :
    QObject(variable.parent())
{
    this->_sendEvents = variable.sendEventsChange();
    this->_name = variable.getAttribute(Name);
    this->_dataType = variable.getAttribute(DataType);
    this->_defaultValue = variable.getAttribute(DefaultValue);
    this->_maximum = variable.getAttribute(Maximum);
    this->_minimum = variable.getAttribute(Minimum);
    this->_step = variable.getAttribute(Step);
    this->_value = variable.getValue();
}

StateVariable &StateVariable::operator=(const StateVariable &variable) {
    if (this != &variable) {
        this->nextMulticastSeq = 0;
        this->setParent(variable.parent());
        this->_sendEvents = variable.sendEventsChange();
        this->_name = variable.getAttribute(Name);
        this->_dataType = variable.getAttribute(DataType);
        this->_defaultValue = variable.getAttribute(DefaultValue);
        this->_maximum = variable.getAttribute(Maximum);
        this->_minimum = variable.getAttribute(Minimum);
        this->_step = variable.getAttribute(Step);
        this->_value = variable.getValue();
        this->_multicast = variable.multicastEvents();
    }
    return *this;
}

void StateVariable::setAttribute(BrisaStateVariableAttribute attr, QVariant newValue) {
    switch (attr) {
    case SendEvents:
    case sendEvents:
        this->_sendEvents = (newValue.toString() == "yes");
        break;
    case Name:
    case name:
        this->_name = newValue.toString();
        break;
    case DataType:
    case dataType:
        this->_dataType = newValue.toString();
        break;
    case DefaultValue:
    case defaultValue:
        this->_defaultValue = newValue.toString();
        break;
    case AllowedValue:
    case allowedValue:
        this->allowedValueList.append(newValue.toString());
        break;
    case Minimum:
    case minimum:
        this->_minimum = newValue.toString();
        break;
    case Maximum:
    case maximum:
        this->_maximum = newValue.toString();
        break;
    case Step:
    case step:
        this->_step = newValue.toString();
        break;
    case Value:
    case value:
        if (this->validateNewValue(newValue) && this->_value != newValue) {
            this->_value = newValue;
            if (this->_sendEvents) {
                emit this->changed(this);
            }
        }
        break;
    case Multicast:
    case multicast:
        this->_multicast = (newValue.toString() == "yes");
        break;
    default:
        break;
    }
}

QString StateVariable::getAttribute(BrisaStateVariableAttribute attr, int ind) const {
    switch (attr) {
    case Name:
    case name:
        return this->_name;
        break;
    case DataType:
    case dataType:
        return this->_dataType;
        break;
    case DefaultValue:
    case defaultValue:
        return this->_defaultValue;
        break;
    case AllowedValue:
    case allowedValue:
        return this->allowedValueList[ind];
        break;
    case Minimum:
    case minimum:
        return QString(this->_minimum);
        break;
    case Maximum:
    case maximum:
        return QString(this->_maximum);
        break;
    case Step:
    case step:
        return QString(this->_step);
        break;
    case Value:
    case value:
        return QString(this->_value.toString());
        break;
    case Multicast:
    case multicast:
        return QString(this->_multicast);
        break;
    default:
        return "";
        break;
    }
}

QList<QString> StateVariable::getAllowedValueList() {
    return this->allowedValueList;
}

void StateVariable::addAllowedValue(QString allowedVal) {
    allowedValueList.append(allowedVal);
}

void StateVariable::clear() {
    this->allowedValueList.clear();
    this->_name.clear();
    this->_dataType.clear();
    this->_defaultValue.clear();
    this->_maximum.clear();
    this->_minimum.clear();
    this->_step.clear();
    this->_value.clear();
}

QVariant StateVariable::getValue() const {
    return this->_value;
}

bool StateVariable::sendEventsChange() const {
    return this->_sendEvents;
}

bool StateVariable::multicastEvents() const {
    return this->_multicast;
}

bool StateVariable::validateNewValue(const QVariant &value) {
    QVariant::Type type = this->getDataType();

    if (!value.canConvert(type)) {
        qDebug() << "Type mismatch while validating new value for state variable " << this->_name;
        return false;
    }

    if (type == QVariant::String) {
        if (this->allowedValueList.isEmpty())
            return true;
        if (this->allowedValueList.contains(value.toString())) {
            return true;
        } else {
            qDebug() << "State variable " << this->_name << " does not allows value " << value.toString();
            return false;
        }

    } else if (type == QVariant::Bool) {
        if (value.toString() == "0" || value.toString() == "1"
                || value.toString() == "true" || value.toString() == "false"
                || value.toString() == "yes" || value.toString() == "no") {
            return true;
        } else {
            qDebug() << "Value not allowed for boolean type. Use '0', '1', 'true', 'false', 'yes', or 'no'.";
            return false;
        }

    } else if (type == QVariant::UInt) {
        if (this->_maximum == "" || this->_minimum == "") {
            return true;
        } else if (value.toUInt() <= this->_maximum.toUInt() && value.toUInt() >= this->_minimum.toUInt()) {
            return true;
        } else {
            qDebug() << "Value out of range for UInt.";
            return false;
        }

    } else if (type == QVariant::Int) {
        if (this->_maximum == "" || this->_minimum == "") {
            return true;
        } else if (value.toInt() <= this->_maximum.toInt() && value.toInt() >= this->_minimum.toInt()) {
            return true;
        } else {
            qDebug() << "Value out of range for Int.";
            return false;
        }

    } else if (type == QVariant::Double) {
        if (this->_maximum == "" || this->_minimum == "") {
            return true;
        } else if (value.toDouble() <= this->_maximum.toDouble() && value.toDouble() >= this->_minimum.toDouble()) {
            return true;
        } else {
            qDebug() << "Value out of range for Double.";
            return false;
        }

    } else {
        return true;
    }
}

QVariant::Type StateVariable::getDataType() const {
    if (this->_dataType == "ui1") {
        return QVariant::UInt;
    } else if (this->_dataType == "ui2") {
        return QVariant::UInt;
    } else if (this->_dataType == "ui4") {
        return QVariant::UInt;
    } else if (this->_dataType == "i1") {
        return QVariant::Int;
    } else if (this->_dataType == "i2") {
        return QVariant::Int;
    } else if (this->_dataType == "i4") {
        return QVariant::Int;
    } else if (this->_dataType == "int") {
        return QVariant::Int;
    } else if (this->_dataType == "r4") {
        return QVariant::Double;
    } else if (this->_dataType == "r8") {
        return QVariant::Double;
    } else if (this->_dataType == "number") {
        return QVariant::Double;
    } else if (this->_dataType == "fixed") {
        return QVariant::Double;
    } else if (this->_dataType == "float") {
        return QVariant::Double;
    } else if (this->_dataType == "char") {
        return QVariant::Char;
    } else if (this->_dataType == "string") {
        return QVariant::String;
    } else if (this->_dataType == "date") {
        return QVariant::Date;
    } else if (this->_dataType == "dateTime") {
        return QVariant::DateTime;
    } else if (this->_dataType == "dateTime.tz") {
        return QVariant::DateTime;
    } else if (this->_dataType == "time") {
        return QVariant::Time;
    } else if (this->_dataType == "time.tz") {
        return QVariant::Time;
    } else if (this->_dataType == "boolean") {
        return QVariant::Bool;
    } else if (this->_dataType == "bin.base64") {
        return QVariant::String;
    } else if (this->_dataType == "bin.hex") {
        return QVariant::String;
    } else if (this->_dataType == "uri") {
        return QVariant::String;
    } else if (this->_dataType == "uuid") {
        return QVariant::String;
    } else {
        return QVariant::Invalid;
    }
}

void StateVariable::setMulticastSeq(quint32 *seq)
{
    this->nextMulticastSeq = seq;
}

quint32 StateVariable::getNextMulticastSeq()
{
    if (*(this->nextMulticastSeq) == 0) {
        (*(this->nextMulticastSeq))++;
        return 0;
    }

    if (*(this->nextMulticastSeq) == 0xFFFFFFFF) /*4294967295L*/
        *(this->nextMulticastSeq) = 0;

    return (*(this->nextMulticastSeq))++;
}

void StateVariable::setMulticastUsn(QString usn)
{
    this->multicastUsn = usn;
}

QString StateVariable::getMulticastUsn()
{
    return this->multicastUsn;
}

void StateVariable::setMulticastUdn(QString udn)
{
    this->multicastUdn = udn;
}

QString StateVariable::getMulticastUdn()
{
    return this->multicastUdn;
}

void StateVariable::setMulticastSvcid(QString svcid)
{
    this->multicastSvcid = svcid;
}

QString StateVariable::getMulticastSvcid()
{
    return this->multicastSvcid;
}

}
}

File: java/trunk/jbrisa/JBRisa/src/jbrisa/upnp/device/model/BrisaStateVariable.java
package jbrisa.upnp.device.model;

import java.lang.reflect.Type;
import java.net.URI;
import java.sql.Time;
import java.util.ArrayList;
import java.util.Date;
import java.util.Observable;
import java.util.UUID;

public class BrisaStateVariable extends Observable {
	
	private String name;
	private String defaultValue;
	private String maximum;
	private String minimum;
	private String step;
	private String dataType;
	private String strSendEvents;
	private String strMulticast;
	
	private ArrayList<String> allowedValueList;
	
	private boolean sendEvents;
	private boolean multicast;
	
	private String value;

	private Integer nextMulticastSeq;
	//private String multicastUsn;
	//private String multicastUdn;
	//private String multicastSvcid;	

	public BrisaStateVariable(boolean sendEvents, String name, String dataType,
			String defaultValue, String maximum, String minimum, String step,
			boolean multicast) {
		this.sendEvents = sendEvents;
		this.name = name;
		this.dataType = dataType;
		this.defaultValue = defaultValue;
		this.maximum = maximum;
		this.minimum = minimum;
		this.step = step;
		this.multicast = multicast;
		
		this.allowedValueList = new ArrayList<String>();
		
		this.nextMulticastSeq = 1;
	}
	
	public BrisaStateVariable() {
		this(false, "", "", "", "", "", "", false);
	}

	public boolean validadeNewValue(Object newValue) {
		//TODO Validar valor
		return true;
	}
	
	
	public void setDataType(String dataType) {
		this.dataType = dataType;
	}

	public String getStrDataType() {
		return dataType;
	}
	
	/**
	 * 
	 * @return
	 */
	public Type getDataType() {
		
		if (this.dataType == "ui1") {
	        return Integer.TYPE;
	    } else if (this.dataType == "ui2") {
	    	return Integer.TYPE;
	    } else if (this.dataType == "ui4") {
	    	return Integer.TYPE;
	    } else if (this.dataType == "i1") {
	    	return int.class;
	    } else if (this.dataType == "i2") {
	    	return Integer.TYPE;
	    } else if (this.dataType == "i4") {
	    	return Integer.TYPE;
	    } else if (this.dataType == "int") {
	    	return Integer.TYPE;
	    } else if (this.dataType == "r4") {
	        return Double.TYPE;
	    } else if (this.dataType == "r8") {
	    	return Double.TYPE;
	    } else if (this.dataType == "number") {
	    	return Double.TYPE;
	    } else if (this.dataType == "fixed") {
	    	return Double.TYPE;
	    } else if (this.dataType == "float") {
	    	return Double.TYPE;
	    } else if (this.dataType == "char") {
	        return Character.TYPE;
	    } else if (this.dataType == "string") {
	        return String.class;
	    } else if (this.dataType == "date") {
	        return Date.class;
	    } else if (this.dataType == "dateTime") {
	        return Date.class;
	    } else if (this.dataType == "dateTime.tz") {
	    	return Date.class;
	    } else if (this.dataType == "time") {
	        return Time.class;
	    } else if (this.dataType == "time.tz") {
	        return Time.class;
	    } else if (this.dataType == "boolean") {
	        return Boolean.TYPE;
	    } else if (this.dataType == "bin.base64") {
	        return Byte.TYPE;
	    } else if (this.dataType == "bin.hex") {
	        return Byte.TYPE;
	    } else if (this.dataType == "uri") {
	        return URI.class;
	    } else if (this.dataType == "uuid") {
	        return UUID.class;
	    } else {
	        return null;
	    }
		
	}
	
	public void setNextMulticastSeq(Integer nextMulticastSeq) {
		this.nextMulticastSeq = nextMulticastSeq;
	}

	/**
	 * O proximo contador do multicast
	 * @return
	 */
	public Integer getNextMulticastSeq() {
		
		if (this.nextMulticastSeq == Integer.MAX_VALUE) 
			this.nextMulticastSeq = 0;

		return this.nextMulticastSeq++;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getDefaultValue() {
		return defaultValue;
	}

	public void setDefaultValue(String defaultValue) {
		this.defaultValue = defaultValue;
	}

	///////////////////////////////////////////////
	public String getMaximum() {
		return maximum;
	}

	public void setMaximum(String maximum) {
		this.maximum = maximum;
	}

	public String getMinimum() {
		return minimum;
	}

	public void setMinimum(String minimum) {
		this.minimum = minimum;
	}

	public String getStep() {
		return step;
	}

	public void setStep(String step) {
		this.step = step;
	}
	
	///////////////////////////////////////////////
	
	public ArrayList<String> getAllowedValueList() {
		return allowedValueList;
	}

	public void setAllowedValueList(ArrayList<String> allowedValueList) {
		this.allowedValueList = allowedValueList;
	}
	
	////////////////////////////////////////////////

//	public void setMulticastUsn(String multicastUsn) {
//		this.multicastUsn = multicastUsn;
//	}
//
//
//	public String getMulticastUdn() {
//		return multicastUdn;
//	}
//
//	public void setMulticastUdn(String multicastUdn) {
//		this.multicastUdn = multicastUdn;
//	}
//
//	public String getMulticastSvcid() {
//		return multicastSvcid;
//	}
//
//	public void setMulticastSvcid(String multicastSvcid) {
//		this.multicastSvcid = multicastSvcid;
//	}

	public String getValue() {
		return value;
	}
	
	//FIXME Validar o value
	public void setValue(String newValue) {
		
		if(this.value != newValue) {
			this.value = newValue;
			if(this.multicastEvents()) {
				setChanged();
				notifyObservers();
			}
		}
	}


///////////////////////////////////////////////
	public String getStrSendEvents() {
		return strSendEvents;
	}
	
	public boolean isSendEvents() {
		return sendEvents;
	}

	public void setSendEvents(boolean sendEvents) {
		this.sendEvents = sendEvents;
		if(this.sendEvents) {
			this.strSendEvents = "yes";
		} else {
			this.strSendEvents = "no";
		}
	}
	
	public void setStrSendEvents(String strSendEvents) {
		if(strSendEvents == null)
			return;
		this.strSendEvents = strSendEvents;
		if(this.strSendEvents.equalsIgnoreCase("yes")) {
			this.sendEvents = true;
		} else {
			this.sendEvents = false;
		}
	}

	///////////////////////////////////////////////////////
	public boolean multicastEvents() {
		return multicast;
	}
	
	public boolean unicastEvents(){
		return sendEvents;
	}

	public void setMulticast(boolean multicast) {
		this.multicast = multicast;
		if(this.multicast) {
			this.strMulticast = "yes";
		} else {
			this.strMulticast = "no";
		}
	}

	/**
	 * 
	 * @param strMulticast
	 */
	public void setStrMulticast(String strMulticast) {
		if(strMulticast == null)
			return;
		this.strMulticast = strMulticast;
		if(this.strMulticast.equalsIgnoreCase("yes")) {
			this.multicast = true;
		} else {
			this.multicast = false;
		}
	}

	public String getStrMulticast() {
		return strMulticast;
	}
	
	
}
